### PR TITLE
update pidcontroller for 2018 some more

### PR DESCRIPTION
**Pull request by ariovistus**
also, add a few unit tests

### CODE REVIEW

File: wpilib/wpilib/pidcontroller.py [hunk 1]
```diff
@@ -141,67 +136,76 @@ def __init__(self, Kp, Ki, Kd, *args, **kwargs):
...
-                self.result = self.clamp(self.result, self.minimumOutput, self.maximumOutput)
+            result = self.clamp(result, minimumOutput, maximumOutput)
 
-                pidOutput = self.pidOutput
-                result = self.result
-                
-                self.buf.append(self.error)
+            with self.pidWriteMutex:
+                with self.mutex:
+                    enabled = self.enabled
```
Comment (ariovistus) [inline wpilib/wpilib/pidcontroller.py]: Need eyes on this bit. It looks to me like this is equivalent to what's in upstream, but not sure.
It also looks like upstream could have just called isEnabled

**Comment by ariovistus** [comment]
question for upstream: reset uses reentrant locks but is still adorned with synchronized. Is it supposed to be this way and if so why?

File: wpilib/wpilib/pidcontroller.py (hunk 1 above)
Comment (virtuald) [reply wpilib/wpilib/pidcontroller.py]: Yeah, this interpretation looks correct to me.

**Comment by virtuald** [comment]
I don't see a synchronized marker anywhere in PIDController?

**Comment by ariovistus** [comment]
hm. must have been looking at de63e1c8a1131b0d23bdc78c3a09896e4037bbbe as presented by source track.

**Comment by auscompgeek** [comment]
I'm not seeing any synchronized markers either. We have 899892c1193fe21afdfb8a6f20b007e4ce609120 currently tracked by git-source-track (although we should bump that).

**Pull request merged**